Migrate DQMEDAnalyze to producers

dmitrijus opened the pull request:
Also fixes MEtoEDMConverter.

Supersedes https://github.com/cms-sw/cmssw/pull/22307

schneiml committed:
Switch to EDProducer.

Also re-introduce the .cc file, implementations in the header are to likely to confuse the build system.

(cherry picked from commit 6ea361376baee5c97bbf91692935585ec0f55a24)

Commit:
Implement DQMToken @ Run support in DQMEDHarvester.

Removes unused meGetter better from DQMStore too.

schneiml committed:
Convert EDMtoMEConverter to a producer, similar to DQMEDHarvester.

Also protect the DQMStore using usesResource, since all of this is
really dangerous.

(cherry picked from commit fdbe8d76cb6ca1db06c19d81bcaadf38b23e6a35)

schneiml committed:
Some formal changes to MEtoEDMConverter.

Consume tokens, protect the DQMStore and watch all the transitions.

(cherry picked from commit ebe0a87bd3ab878194533726b48b032915adf413)

schneiml committed:
Another DQMEDAnalyzer config change

(cherry picked from commit 85a9c26fbf7073ea83da6ef06878d497289cac35)

Commit:
Copy run histograms as well as lumi histograms.

Commit:
Modernize MEtoEDMConverter.

1. Remove useless statistics
2. Remove run reset (it does nothing multithread)
3. Switch to use iBooker/iGetter.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22319/3528

cmsbuild commented:
A new Pull Request was created by  @dmitrijus (Dmitrijus) for master.

It involves the following packages:

DQMServices/Components
DQMServices/Core
DataFormats/Histograms
Validation/RecoTau


@smuzaffar, @Dr15Jones, @vazzolini, @kmaeshima, @dmitrijus, @cmsbuild, @jfernan2, @vanbesien can you please review it and eventually sign? Thanks.
@barvic, @rovere this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

davidlange6 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

Commit:
Modernize EDMtoMEConverter. Should not change behaviour in any way.

Commit:
Remove unused DQMStore::getAllTags.

cmsbuild commented:
-1

Tested at: c528197348baadf60dd19b3f8414f6f8c49bee51

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22319/26265/summary.html

I found follow errors while testing this PR

**Failed tests:  Build**

* **Build**:

I found an error when building: 
 
 <pre>                 from /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_1_X_2018-02-23-1100/src/DQM/EcalPreshowerMonitorClient/src/EcalPreshowerMonitorClient.cc:1:
/cvmfs/cms-ib.cern.ch/nweek-02512/slc6_amd64_gcc630/external/gcc/6.3.0-cms/include/c++/6.3.0/bits/unique_ptr.h: In instantiation of 'typename std::_MakeUniq<_Tp>::__single_object std::make_unique(_Args&& ...) [with _Tp = EcalPreshowerMonitorClient; _Args = {const edm::ParameterSet&}; typename std::_MakeUniq<_Tp>::__single_object = std::unique_ptr<EcalPreshowerMonitorClient, std::default_delete<EcalPreshowerMonitorClient> >]':
/cvmfs/cms-ib.cern.ch/nweek-02512/slc6_amd64_gcc630/cms/cmssw-patch/CMSSW_10_1_X_2018-02-23-1100/src/FWCore/Framework/src/MakeModuleHelper.h:41:40:   required from 'static std::unique_ptr<_Tp> edm::MakeModuleHelper<Base>::makeModule(const edm::ParameterSet&) [with T = EcalPreshowerMonitorClient; Base = edm::one::EDProducerBase]'
/cvmfs/cms-ib.cern.ch/nweek-02512/slc6_amd64_gcc630/cms/cmssw-patch/CMSSW_10_1_X_2018-02-23-1100/src/FWCore/Framework/src/WorkerMaker.h:85:147:   required from 'std::shared_ptr<edm::maker::ModuleHolder> edm::WorkerMaker<T>::makeModule(const edm::ParameterSet&) const [with T = EcalPreshowerMonitorClient]'
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_1_X_2018-02-23-1100/src/DQM/EcalPreshowerMonitorClient/src/EcalPreshowerMonitorClient.cc:103:46:   required from here
/cvmfs/cms-ib.cern.ch/nweek-02512/slc6_amd64_gcc630/external/gcc/6.3.0-cms/include/c++/6.3.0/bits/unique_ptr.h:791:30: error: invalid new-expression of abstract class type 'EcalPreshowerMonitorClient'
     { return unique_ptr<_Tp>(new _Tp(std::forward<_Args>(__args)...)); }
                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_1_X_2018-02-23-1100/src/DQM/EcalPreshowerMonitorClient/src/EcalPreshowerMonitorClient.cc:1:0:
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_1_X_2018-02-23-1100/src/DQM/EcalPreshowerMonitorClient/interface/EcalPreshowerMonitorClient.h:8:7: note:   because the following virtual functions are pure within 'EcalPreshowerMonitorClient':
 class EcalPreshowerMonitorClient : public DQMEDHarvester {
</pre>

cmsbuild commented:
Comparison not run due to Build errors (RelVals and Igprof tests were also skipped)

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22319/3534

Code check has found code style and quality issues which could be resolved by applying a patch in https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22319/3534/git-diff.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22319/3534/git-diff.patch | patch -p1`

You can run `scram build code-checks` to apply code checks directly

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22319/3537

Code check has found code style and quality issues which could be resolved by applying a patch in https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22319/3537/git-diff.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22319/3537/git-diff.patch | patch -p1`

You can run `scram build code-checks` to apply code checks directly

Commit:
Apply code-checks.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22319/3539

cmsbuild commented:
Pull request #22319 was updated. @smuzaffar, @Dr15Jones, @vazzolini, @kmaeshima, @dmitrijus, @cmsbuild, @jfernan2, @vanbesien can you please check and sign again.

dmitrijus commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

fwyzard commented on DQMServices/Components/plugins/EDMtoMEConverter.h:
```diff
@@ -56,15 +56,16 @@ class EDMtoMEConverter : public edm::EDAnalyzer
   void beginLuminosityBlock(const edm::LuminosityBlock&, const edm::EventSetup&) override;
   void endLuminosityBlock(const edm::LuminosityBlock&, const edm::EventSetup&) override;
-  void respondToOpenInputFile(const edm::FileBlock&) override;
+  void respondToOpenInputFile(const edm::FileBlock&) ;
```
if this is never called, you could simply drop it, as well as `iCountf`

dmitrijus replied:
will be removed.

fwyzard commented on DQMServices/Core/interface/DQMEDAnalyzer.h:
```diff
@@ -6,58 +6,39 @@
   DQMEDAnalyzer(DQMEDAnalyzer const&) = delete;
   DQMEDAnalyzer(DQMEDAnalyzer &&) = delete;
 
-  void beginRun(edm::Run const& run, edm::EventSetup const& setup) final
```
same comment as in #22307: keep the trivial methods in the .h file

dmitrijus commented:
@Dr15Jones I am using:

```
produces<DQMToken,edm::Transition::EndLuminosityBlock>();
produces<DQMToken,edm::Transition::EndRun>();
```

This gives me:

```
----- Begin Fatal Exception 23-Feb-2018 18:17:09 CET-----------------------
An exception of category 'LogicError' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=TkAlCaRecoMonitor label='ALCARECOSiStripCalMinBiasAAGTrackerDQM'
Exception Message:
Duplicate Product The produced product Branch Type = LuminosityBlock
Process Name = RECO
ModuleLabel = ALCARECOSiStripCalMinBiasAAGTrackerDQM
Branch ID = 1450585091
Class Name = DQMToken
Friendly Class Name = DQMToken
Product Instance Name = 
 is registered more than once.
Please remove the redundant 'produces' call(s).
----- End Fatal Exception -------------------------------------------------
```

Should I go back to using two tokens: DQMToken and DQMRunToken? The code works fine with those.

If I give them labels, it works too:
```
produces<DQMToken,edm::Transition::EndLuminosityBlock>("endLumi");
produces<DQMToken,edm::Transition::EndRun>("endRun");
```

Can you confirm that consumesAll works fine with different labels?

Dr15Jones commented:
@dmitrijus we think we know why you get the unexpected behavior. When telling an OutputModule what to `keep`, there is no way to specify if the item is in the Run or LuminosityBlock. So I believe the framework is requiring a unit branch name across Runs, LuminosityBlocks and Events. Not good, but that is what we have.

>Can you confirm that consumesAll works fine with different labels?

That is how the framework should work. I do not know if we've tested that explicitly.

Commit:
Use labels for DQMToken.

Also: remove useless private statistics from EDMtoMEConverter
Also: fix EDMtoMEConverter to actually produce DQMTokens

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22319/3542

cmsbuild commented:
-1

Tested at: 5424ec4e802363f9489000162f9b3032649743d1

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22319/26270/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests  RelVals  AddOn**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test testSSTGainPCL_fromRECO had ERRORS


* **RelVals**:

When I ran the RelVals I found an error in the following worklfows:
4.22 step2
<pre>runTheMatrix-results/4.22_RunCosmics2011A+RunCosmics2011A+RECOCOSD+ALCACOSD+SKIMCOSD+HARVESTDC/step2_RunCosmics2011A+RunCosmics2011A+RECOCOSD+ALCACOSD+SKIMCOSD+HARVESTDC.log

</pre>5.1 step1
<pre>runTheMatrix-results/5.1_TTbar+TTbarFS+HARVESTFS/step1_TTbar+TTbarFS+HARVESTFS.log

</pre>135.4 step1
<pre>runTheMatrix-results/135.4_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS/step1_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS.log

</pre>8.0 step3
<pre>runTheMatrix-results/8.0_BeamHalo+BeamHalo+DIGICOS+RECOCOS+ALCABH+HARVESTCOS/step3_BeamHalo+BeamHalo+DIGICOS+RECOCOS+ALCABH+HARVESTCOS.log

</pre>4.53 step3
<pre>runTheMatrix-results/4.53_RunPhoton2012B+RunPhoton2012B+HLTD+RECODR1reHLT+HARVESTDR1reHLT/step3_RunPhoton2012B+RunPhoton2012B+HLTD+RECODR1reHLT+HARVESTDR1reHLT.log

</pre>7.3 step3
<pre>runTheMatrix-results/7.3_CosmicsSPLoose_UP17+CosmicsSPLoose_UP17+DIGICOS_UP17+RECOCOS_UP17+ALCACOS_UP17+HARVESTCOS_UP17/step3_CosmicsSPLoose_UP17+CosmicsSPLoose_UP17+DIGICOS_UP17+RECOCOS_UP17+ALCACOS_UP17+HARVESTCOS_UP17.log

</pre>136.7611 step2
<pre>runTheMatrix-results/136.7611_RunJetHT2016E_reminiaod+RunJetHT2016E_reminiaod+REMINIAOD_data2016_HIPM+HARVESTDR2_REMINIAOD_data2016_HIPM/step2_RunJetHT2016E_reminiaod+RunJetHT2016E_reminiaod+REMINIAOD_data2016_HIPM+HARVESTDR2_REMINIAOD_data2016_HIPM.log

</pre>140.53 step2
<pre>runTheMatrix-results/140.53_RunHI2011+RunHI2011+RECOHID11+HARVESTDHI/step2_RunHI2011+RunHI2011+RECOHID11+HARVESTDHI.log

</pre>136.731 step3
<pre>runTheMatrix-results/136.731_RunSinglePh2016B+RunSinglePh2016B+HLTDR2_2016+RECODR2_2016reHLT_skimSinglePh_HIPM+HARVESTDR2/step3_RunSinglePh2016B+RunSinglePh2016B+HLTDR2_2016+RECODR2_2016reHLT_skimSinglePh_HIPM+HARVESTDR2.log

</pre>136.788 step3
<pre>runTheMatrix-results/136.788_RunSinglePh2017B+RunSinglePh2017B+HLTDR2_2017+RECODR2_2017reHLT_skimSinglePh_Prompt+HARVEST2017/step3_RunSinglePh2017B+RunSinglePh2017B+HLTDR2_2017+RECODR2_2017reHLT_skimSinglePh_Prompt+HARVEST2017.log

</pre>1325.7 step2
<pre>runTheMatrix-results/1325.7_TTbar_13_94XNanoAODINPUT+TTbar_13_94XNanoAODINPUT+NANOEDMMC2017+HARVESTNANOAODMC2017/step2_TTbar_13_94XNanoAODINPUT+TTbar_13_94XNanoAODINPUT+NANOEDMMC2017+HARVESTNANOAODMC2017.log

</pre>101.0 step1
<pre>runTheMatrix-results/101.0_SingleElectronE120EHCAL+SingleElectronE120EHCAL/step1_SingleElectronE120EHCAL+SingleElectronE120EHCAL.log

</pre>9.0 step3
<pre>runTheMatrix-results/9.0_Higgs200ChargedTaus+Higgs200ChargedTaus+DIGI+RECO+HARVEST/step3_Higgs200ChargedTaus+Higgs200ChargedTaus+DIGI+RECO+HARVEST.log

</pre>1306.0 step3
<pre>runTheMatrix-results/1306.0_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15/step3_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15.log

</pre>1001.0 step2
<pre>runTheMatrix-results/1001.0_RunMinBias2011A+RunMinBias2011A+TIER0EXP+ALCAEXP+ALCAHARVD1+ALCAHARVD2+ALCAHARVD3+ALCAHARVD4+ALCAHARVD5/step2_RunMinBias2011A+RunMinBias2011A+TIER0EXP+ALCAEXP+ALCAHARVD1+ALCAHARVD2+ALCAHARVD3+ALCAHARVD4+ALCAHARVD5.log

</pre>1000.0 step2
<pre>runTheMatrix-results/1000.0_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT/step2_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT.log

</pre>25.0 step3
<pre>runTheMatrix-results/25.0_TTbar+TTbar+DIGI+RECOAlCaCalo+HARVEST+ALCATT/step3_TTbar+TTbar+DIGI+RECOAlCaCalo+HARVEST+ALCATT.log

</pre>150.0 step3
<pre>runTheMatrix-results/150.0_HydjetQ_B12_5020GeV_2018+HydjetQ_B12_5020GeV_2018+DIGIHI2018+RECOHI2018+HARVESTHI2018/step3_HydjetQ_B12_5020GeV_2018+HydjetQ_B12_5020GeV_2018+DIGIHI2018+RECOHI2018+HARVESTHI2018.log

</pre>1330.0 step3
<pre>runTheMatrix-results/1330.0_ZMM_13+ZMM_13+DIGIUP15+RECOUP15_L1TMuDQM+HARVESTUP15_L1TMuDQM/step3_ZMM_13+ZMM_13+DIGIUP15+RECOUP15_L1TMuDQM+HARVESTUP15_L1TMuDQM.log

</pre>10042.0 step3
<pre>runTheMatrix-results/10042.0_ZMM_13+ZMM_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017/step3_ZMM_13+ZMM_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017.log

</pre>25202.0 step3
<pre>runTheMatrix-results/25202.0_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25/step3_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25.log

</pre>10024.0 step3
<pre>runTheMatrix-results/10024.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017/step3_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017.log

</pre>11624.0 step3
<pre>runTheMatrix-results/11624.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2019_GenSimFull+DigiFull_2019+RecoFull_2019+ALCAFull_2019+HARVESTFull_2019/step3_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2019_GenSimFull+DigiFull_2019+RecoFull_2019+ALCAFull_2019+HARVESTFull_2019.log

</pre>10824.0 step3
<pre>runTheMatrix-results/10824.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2018_GenSimFull+DigiFull_2018+RecoFull_2018+ALCAFull_2018+HARVESTFull_2018/step3_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2018_GenSimFull+DigiFull_2018+RecoFull_2018+ALCAFull_2018+HARVESTFull_2018.log

</pre>10224.0 step3
<pre>runTheMatrix-results/10224.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017PU_GenSimFull+DigiFullPU_2017PU+RecoFullPU_2017PU+HARVESTFullPU_2017PU/step3_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017PU_GenSimFull+DigiFullPU_2017PU+RecoFullPU_2017PU+HARVESTFullPU_2017PU.log

</pre>20034.0 step3
<pre>runTheMatrix-results/20034.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D17_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D17+RecoFullGlobal_2023D17+HARVESTFullGlobal_2023D17/step3_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D17_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D17+RecoFullGlobal_2023D17+HARVESTFullGlobal_2023D17.log

</pre>21234.0 step3
<pre>runTheMatrix-results/21234.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D21_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D21+RecoFullGlobal_2023D21+HARVESTFullGlobal_2023D21/step3_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D21_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D21+RecoFullGlobal_2023D21+HARVESTFullGlobal_2023D21.log

</pre>20434.0 step3
<pre>runTheMatrix-results/20434.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D19_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D19+RecoFullGlobal_2023D19+HARVESTFullGlobal_2023D19/step3_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D19_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D19+RecoFullGlobal_2023D19+HARVESTFullGlobal_2023D19.log

</pre>

* **AddOn**:

I found errors in the following addon tests:

cmsDriver.py TTbar_8TeV_TuneCUETP8M1_cfi  --conditions auto:run1_mc --fast  -n 100 --eventcontent AODSIM,DQM --relval 100000,1000 -s GEN,SIM,RECOBEFMIX,DIGI:pdigi_valid,L1,DIGI2RAW,L1Reco,RECO,EI,VALIDATION  --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --datatier GEN-SIM-DIGI-RECO,DQMIO --beamspot Realistic8TeVCollision : FAILED - time: date Fri Feb 23 19:36:09 2018-date Fri Feb 23 19:35:04 2018 s - exit: 18944
cmsDriver.py TTbar_13TeV_TuneCUETP8M1_cfi --conditions auto:run2_mc_l1stage1 --fast  -n 100 --eventcontent AODSIM,DQM --relval 100000,1000 -s GEN,SIM,RECOBEFMIX,DIGI:pdigi_valid,L1,DIGI2RAW,L1Reco,RECO,EI,VALIDATION  --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --datatier GEN-SIM-DIGI-RECO,DQMIO --beamspot NominalCollision2015 --era Run2_25ns : FAILED - time: date Fri Feb 23 19:36:35 2018-date Fri Feb 23 19:35:23 2018 s - exit: 18944
cmsDriver.py TTbar_13TeV_TuneCUETP8M1_cfi --conditions auto:run2_mc --fast  -n 100 --eventcontent AODSIM,DQM --relval 100000,1000 -s GEN,SIM,RECOBEFMIX,DIGI:pdigi_valid,L1,DIGI2RAW,L1Reco,RECO,EI,VALIDATION  --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --datatier GEN-SIM-DIGI-RECO,DQMIO --beamspot NominalCollision2015 --era Run2_2016 : FAILED - time: date Fri Feb 23 19:37:48 2018-date Fri Feb 23 19:35:59 2018 s - exit: 18944

cmsbuild commented:
Pull request #22319 was updated. @smuzaffar, @Dr15Jones, @vazzolini, @kmaeshima, @dmitrijus, @cmsbuild, @jfernan2, @vanbesien can you please check and sign again.

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

dmitrijus commented:
@Dr15Jones I have done the labels in the https://github.com/cms-sw/cmssw/pull/22319/commits/c27f399a3ed38527f897e585cade081bebd1b839

Should this PR should be ready to be reviewed and merged :)

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: c27f399a3ed38527f897e585cade081bebd1b839

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22319/26278/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals**

* **RelVals**:

When I ran the RelVals I found an error in the following worklfows:
20034.0 step3
<pre>runTheMatrix-results/20034.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D17_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D17+RecoFullGlobal_2023D17+HARVESTFullGlobal_2023D17/step3_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D17_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D17+RecoFullGlobal_2023D17+HARVESTFullGlobal_2023D17.log

</pre>20434.0 step3
<pre>runTheMatrix-results/20434.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D19_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D19+RecoFullGlobal_2023D19+HARVESTFullGlobal_2023D19/step3_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D19_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D19+RecoFullGlobal_2023D19+HARVESTFullGlobal_2023D19.log

</pre>21234.0 step3
<pre>runTheMatrix-results/21234.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D21_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D21+RecoFullGlobal_2023D21+HARVESTFullGlobal_2023D21/step3_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D21_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D21+RecoFullGlobal_2023D21+HARVESTFullGlobal_2023D21.log

</pre>

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

Dr15Jones commented:
The RelVal failures are do to a configuration error
```
----- Begin Fatal Exception 23-Feb-2018 21:05:57 CET-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Validating configuration of module: class=OuterTrackerMonitorTTCluster label='OuterTrackerMonitorTTCluster'
Exception Message:
The base type in the python configuration is EDAnalyzer, but the base type
for the module's C++ class is EDProducer. Please fix the configuration.
It must use the same base type as the C++ class.
----- End Fatal Exception -------------------------------------------------
```

Commit:
Migrate DQM/Phase2OuterTracker to use DQMEDAnalyzer alias.

dmitrijus commented:
please test

dmitrijus commented:
Fixed the issue and rebased onto the latest IB.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22319/3553

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #22319 was updated. @smuzaffar, @Dr15Jones, @vazzolini, @kmaeshima, @dmitrijus, @cmsbuild, @jfernan2, @vanbesien can you please check and sign again.

cmsbuild commented:
+1
Tested at: a897e113803e91810010e8453f390c5fd9e0b7ae
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22319/26286/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22319/26286/summary.html

Comparison Summary:
 - You potentially added 1053 lines to the logs
 - Reco comparison results: 53 differences found in the comparisons
 - DQMHistoTests: Total files compared: 28
 - DQMHistoTests: Total histograms compared: 2477704
 - DQMHistoTests: Total failures: 75929
 - DQMHistoTests: Total nulls: 2
 - DQMHistoTests: Total successes: 2401597
 - DQMHistoTests: Total skipped: 176
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0800000000163 KiB( 4 files compared)
 - Checked 115 log files, 9 edm output root files, 28 DQM output files

dmitrijus commented:
+1

fwyzard commented on DQMServices/Components/plugins/MEtoEDMConverter.cc:
```diff
@@ -92,274 +91,42 @@ void
 {
   // Determine if we are running multithreading asking to the DQMStore. Not to be moved in the ctor
+  DQMStore *dbe = edm::Service<DQMStore>().operator->();
   enableMultiThread_ = dbe->enableMultiThread_;
```
isn't `enableMultiThread_` something you wanted to get rid of ?

dmitrijus replied:
It is. And we are getting rid of it. This PR is a step towards that - or rather to fix the https://github.com/cms-sw/cmssw/issues/22281 with least amount of changes.

fwyzard commented on DQMServices/Core/interface/DQMEDHarvester.h:
```diff
@@ -1,43 +1,45 @@
-class DQMEDHarvester
-: public edm::one::EDProducer<edm::one::WatchRuns,edm::one::WatchLuminosityBlocks,edm::one::SharedResources,
-edm::EndLuminosityBlockProducer>
+class DQMEDHarvester: public edm::one::EDProducer<edm::Accumulator,
```
why do you inherit from `edm::Accumulator`, if you do not read any event data ?

dmitrijus replied:
The module has to implement either accumulate() or produce(), if none is implemented, it gives a compiler error.

Accumulate makes more sense.

fwyzard replied:
My understanding of what the framework will do is:
  - if standard `EDProducer` is not explicitly or implicitly scheduled, its `produce` method will not be called;
  - if an `EDProducer<Accumulator>` is not explicitly or implicitly scheduled, its `accumulate` method will be called for each event.

@Dr15Jones , sorry for bothering you, but did I understand correctly ?

Then, using a standard `EDProducer` for a harvesting module should be (slightly) more efficient.

Dr15Jones replied:
@fwyzard you are correct, the standard produce mechanism would be more efficient.

fwyzard commented on DQMServices/Core/src/DQMEDHarvester.cc:
```diff
@@ -8,24 +8,29 @@
 
 void DQMEDHarvester::endLuminosityBlockProduce(edm::LuminosityBlock& iLumi, edm::EventSetup const&) {
-  iLumi.put(std::make_unique<DQMToken>());
+  iLumi.put(lumiToken_, std::make_unique<DQMToken>());
```
@Dr15Jones I'm not familiar with `EDPutToken`s... is it just an optimisation to use them, or do they enable any different behaviour ?

Dr15Jones replied:
It is an optimization like `EDGetTokens`

dmitrijus commented on DQMServices/Core/src/DQMStore.cc:
```diff
@@ -2070,6 +2025,43 @@ DQMStore::cloneLumiHistograms(uint32_t run, uint32_t lumi, uint32_t moduleId)
+ */
+
+void
+DQMStore::cloneRunHistograms(uint32_t run, uint32_t moduleId)
```
The goal, again, was to restore the previous behaviour (pre-one module): all histograms are cloned.

Ideally, once we get rid of enableMultithread_ stuff, we can get rid away of this copying.

schneiml replied:
#22307 now has re-written query logic, that should work without cloning.

However, I am not that sure if it does the right thing (by default MEs are per-run, but they can also be "global"), while maybe we want to have them global by default and only have per-run ones when needed. It does fix #22281 (even tough some of the changes of #22319 are missing over there), and also works in a number of other sequences, but might break things that I did not test.

Commit:
Move histogram production out of a activity registry callback and into *Produce.

dmitrijus commented on DQMServices/Core/src/DQMStore.cc:
```diff
@@ -395,6 +395,13 @@ DQMStore::DQMStore(const edm::ParameterSet &pset, edm::ActivityRegistry& ar)
         gc.luminosityBlockID().luminosityBlock(),
         mcc.moduleDescription()->id());
   });
```
Good point and already changed.

cmsbuild commented:
The code-checks are being triggered in jenkins.

dmitrijus commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22319/3575

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 2b489f39fda3d9af2150540a58a96ce0c446b620
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22319/26331/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22319/26331/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-22319/136.8311_RunJetHT2017F_reminiaod+RunJetHT2017F_reminiaod+REMINIAOD_data2017+HARVEST2017_REMINIAOD_data2017

Comparison Summary:
 - You potentially added 1020 lines to the logs
 - Reco comparison results: 54 differences found in the comparisons
 - DQMHistoTests: Total files compared: 29
 - DQMHistoTests: Total histograms compared: 2479896
 - DQMHistoTests: Total failures: 75929
 - DQMHistoTests: Total nulls: 2
 - DQMHistoTests: Total successes: 2403789
 - DQMHistoTests: Total skipped: 176
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.120000000003 KiB( 6 files compared)
 - Checked 118 log files, 9 edm output root files, 29 DQM output files

dmitrijus commented:
+1

Dr15Jones commented:
+1
removal of blank link in DataFormats/Histograms

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

knoepfel commented:
@dmitrijus: congrats on much cleaner code--it is much clearer in terms of intent.  

I'm working on the concurrent booking of histograms in `DQMStore`, so I'd like to avoid collisions with you.  So far, so good, I think.  However, to support concurrent booking, the `std::set<MonitorElement>` data member will need to be replaced with something else that does not support the concept of `lower_bound`.  I'm not particularly concerned with that, but I'm giving you a heads-up.

fwyzard commented:
Ha ha ha

fwyzard commented:
(try not to break the DQM GUI, too)

Dr15Jones commented:
@fwyzard see #18574

fwyzard commented:
I would say that
  - we should not have two (groups of) people try to migrate the same code base at the same time
  - however the migration is done, it should be made sure that it does not break the use of the DQM code base outside of CMSSW

@dmitrijus and @schneiml should be working on a test to outline the set of necessary changes, possible steps, and required features.
I would suggest that @knoepfel discuss the possible changes with them beforehand ...

schneiml commented:
@knoepfel yes, we should for sure have a talk at some point. We have a few plans on how to continue development -- and we can use any help we get.

knoepfel commented:
@fwyzard: agreed--let's not collide/duplicate efforts.  

@schneiml, @dmitrijus: I can meet by Zoom, Skype, or whatever.  I can be available as early as 9 am (CST) tomorrow morning (2/28).  Otherwise, it will likely need to wait until next week.

fabiocos commented:
+1
